x-pack/plugins/alerting/server/saved_objects/migrations/8.7/index.ts

```diff
@@ -70,9 +70,29 @@ function addLogViewRefToLogThresholdRule(
   return doc;
 }
 
+function migrateOutcomeOrder(
+  doc: SavedObjectUnsanitizedDoc<RawRule>
+): SavedObjectUnsanitizedDoc<RawRule> {
+  if (!doc.attributes.lastRun) {
+    return doc;
+  }
+
+  const outcome = doc.attributes.lastRun.outcome;
+  return {
+    ...doc,
+    attributes: {
+      ...doc.attributes,
+      lastRun: {
+        ...doc.attributes.lastRun,
+        outcomeOrder: RuleLastRunOutcomeOrderMap[outcome],
```

Comment:
I don't recall off hand, but are there any old legacy `outcome`'s that a user could have lingering around that wouldn't be included in the `RuleLastRunOutcomeOrderMap`? If so, is `undefined` fine here until the next rule execution?

Reply:
The `outcome` field was added in 8.6 with only three values specified in the `RuleLastRunOutcomeOrderMap`. So I don't expect any other values there. That said, `undefined` should also work just fine.

Reply:
Unless I miss something, `outcome` cannot be `undefined`. See the 8.6 migration, for example:
https://github.com/elastic/kibana/blob/2d675b47a7c3210a1b430e89ff51c9d7485d1fd2/x-pack/plugins/alerting/server/saved_objects/migrations/8.6/index.ts#L22-L35

If the `outcome` is unknown, the whole `lastRun` structure is omitted.

Furthermore, adding a new outcome value will cause a type error if the value is not added to the `outcome->order` map:

![Screenshot 2023-02-07 at 10 12 49](https://user-images.githubusercontent.com/1938181/217201766-7647c35f-3a4d-4d81-867b-780257f18c9b.png)

So TypeScript protects us from forgetting to add values to the linked structures. 

That said, if you feel strongly about changing the type of `outcomeOrder` to `number | undefined`, I'm okay with that. It doesn't affect the code of Security Solution in any way, as we don't work with that field directly.

Reply:
Awesome!  We've become more sensitive to migrations recently so wanted to understand this one well.

The main thing I'm worrying about is some bug where we would unintentionally have `outcome` set to some invalid value, as part of normal (presumably buggy!) operation.  And then the migration would end up writing out an `undefined` for the order.  But ... that's valid for that field :-)

So, even in that case, I don't think there's anything else we can do to improve the migration.

The rest of the type-checking on outcome looks good, thanks for explaining.